File: packages/claude-code-toolbox/src/agents/runtime/protocols/handoff.ts
/**
 * Hand-off protocol — each agent decides who (if anyone) takes the next turn.
 *
 * The agent system-prompt wrapper instructs them to end with a `HANDOFF: name`
 * line (or `HANDOFF: done` to finish). We parse that line; if the named agent
 * exists in the team roster, we dispatch to them next, otherwise we stop.
 * This is a custom runtime because it requires inter-turn control flow that
 * native Task can't express as a protocol.
 */
import type { Protocol } from "../runtimeTypes";
import { emitMessage, findAgentByName, makeTotals, runOneTurn } from "./shared";

const HANDOFF_RE = /handoff:\s*([A-Za-z0-9._-]+)\s*$/im;

export const handoff: Protocol = async (ctx) => {
  const totals = makeTotals();
  const roster = ctx.team.agents
    .map((n) => findAgentByName(ctx.agents, n))
    .filter((a): a is NonNullable<typeof a> => a !== undefined);
  if (!roster.length) {
    return { status: "error", totals };
  }
  let current = roster[0];
  let lastMessage = "";
  const max = ctx.team.maxTurns || 12;
  for (let turn = 1; turn <= max && !ctx.signal.aborted; turn++) {
    const prompt = [
      `Team hand-off turn ${turn}/${max}. Team task:`,
      "",
      ctx.userPrompt,
      "",
      lastMessage ? `Previous agent said:\n${lastMessage}` : "(you speak first)",
      "",
      `Available teammates: ${roster.map((a) => a.name).filter((n) => n !== current.name).join(", ") || "(none)"}`,
      "",
      `You are "${current.name}". After your response, end with exactly one line:`,
      `HANDOFF: <teammate-name>  (to pass control)`,
      `HANDOFF: done             (when the task is finished)`,
    ].join("\n");
    const { text, errored } = await runOneTurn({
      spawn: ctx.spawnAgentTurn,
      agent: current,
      prompt,
      runId: ctx.runId,
      turn,
      phase: "none",
      bus: ctx.bus,
      totals,
      cwd: ctx.cwd,
      signal: ctx.signal,
      claudeBin: ctx.claudeBin,
    });
    if (errored) {
      return { status: "error", totals };
    }
    lastMessage = text;
    const match = text.match(HANDOFF_RE);
    if (!match || match[1].toLowerCase() === "done") {
      break;
    }
    const nextAgent = findAgentByName(roster, match[1]);
    if (!nextAgent || nextAgent.name === current.name) {
      break;
    }
    emitMessage(ctx.bus, current.name, nextAgent.name, text);
    current = nextAgent;
  }
  return { status: ctx.signal.aborted ? "aborted" : "completed", totals };
};

File: packages/claude-code-toolbox-intellij/src/main/kotlin/com/amitchorasiya/claude/toolbox/intellij/agents/runtime/protocols/Handoff.kt
package com.amitchorasiya.claude.toolbox.intellij.agents.runtime.protocols

import com.amitchorasiya.claude.toolbox.intellij.agents.runtime.*

private val HANDOFF_RE = Regex("HANDOFF\\s*->\\s*(\\S+)", RegexOption.IGNORE_CASE)

fun handoff(ctx: ProtocolContext): ProtocolResult {
    val totals = RunUsage()
    val maxTurns = ctx.team.maxTurns
    val roster = ctx.agents
    if (roster.isEmpty()) return ProtocolResult("error", totals)

    var current = roster[0]
    var prompt = "Team task: ${ctx.userPrompt}\n\nYou are \"${current.name}\". Respond. If you need another agent, end your reply with HANDOFF -> <agent-name>."

    for (turn in 1..maxTurns) {
        if (ctx.abortFlag.get()) return ProtocolResult("aborted", totals)
        val result = spawnAgentTurn(current, prompt, ctx.runId, turn, "none", ctx.bus, totals, ctx.cwd, ctx.claudeBin, ctx.abortFlag)
        if (result.aborted) return ProtocolResult("aborted", totals)
        val m = HANDOFF_RE.find(result.text)
        if (m == null) break
        val nextName = m.groupValues[1]
        val next = findAgentByName(roster, nextName) ?: break
        emitMessage(ctx.bus, current.name, next.name, "HANDOFF -> ${next.name}")
        current = next
        prompt = "Team task: ${ctx.userPrompt}\n\nPrevious agent (${m.groupValues[0]}) handed off to you.\n\nYou are \"${current.name}\". Continue the work."
    }
    return ProtocolResult("completed", totals)
}
